dir_file() does not work as documented

c-alpha opened the pull request:
The documentation claims it returned ".git/config", but it actually
returns ".gitconfig". This is particularly inconvenient when doing a
read/modify/write of one and the same configuration in place. The
filename parameter to set() is required, but the only way to get the
name of the config file which was read before (the dir_file() method)
returns the wrong file name.

c-alpha committed:
dir_file() does not work as documented

The documentation claims it returned ".git/config", but it actually
returns ".gitconfig". This is particularly inconvenient when doing a
read/modify/write of one and the same configuration in place. The
filename parameter to set() is required, but the only way to get the
name of the config file which was read before (the dir_file() method)
returns the wrong file name.

alexmv commented:
100% agree that the documentation and the code disagree, so something needs to be fixed here.  But  `dir_file` is actually a little complicated to pin down.  Because `Config::GitLike::Git` overrides the `load_dirs` method, it could have found it in `.git/config` -- or just `config`, if the repo is bare.  So you're not guaranteed to be able to update a file in-place if `dir_file` always returns `.git/config`.

Does `is_git_dir`s undocumented behavior of returning the path containing a valid `config` help you here?  One option would be to have `dir_file` return `.git/config` if called without arguments, and if called with a path, call into `is_git_dir` to find an actual accurate path.

c-alpha commented:
> 100% agree that the documentation and the code disagree, so something needs to be fixed here. But `dir_file` is actually a little complicated to pin down. Because `Config::GitLike::Git` overrides the `load_dirs` method, it could have found it in `.git/config` -- or just `config`, if the repo is bare. So you're not guaranteed to be able to update a file in-place if `dir_file` always returns `.git/config`.

Good point; I hadn't thought about the case of a bare repo.

> Does `is_git_dir`s undocumented behavior of returning the path containing a valid `config` help you here?

Hidden goodness going on here... 😉 That would be exactly what I need.

> One option would be to have `dir_file` return `.git/config` if called without arguments, and if called with a path, call into `is_git_dir` to find an actual accurate path.

Why not just always call `is_git_dir` from `dir_file` then?

I have experimented a bit, and it seems `is_git_dir` returns the directory where the git files are (i.e. `.../.git` for a working copy, the current directory for a bare repo, and `undef` outside a git tree). Thus, `/config` would need to be appended when the result is not `undef`. 

This led me to the following in my script:
```perl
no warnings qw(redefine);
sub Config::GitLike::Git::dir_file {
    my $self = shift;
    my $path = $self->is_git_dir(".");
    if ((defined $path) and ($path ne "")) {
	my($vol, $dirs, undef) = File::Spec->splitpath( $path, 1 );
	my @dirs               = File::Spec->splitdir( $dirs );
	$path = File::Spec->catfile( @dirs, "config");
    }
    return $path;
}
use warnings qw(redefine);
```
Which seems to do the job (for my purposes at least).

If `dir_file` were changed to accept an argument, that would be an API change.

c-alpha committed:
Infer dir_file Dynamically

Depending on whether a repo is a working copy or a bare repo, the
per-directory git config's path-name is different. Using undocumented,
hidden goodness of is_git_dir to figure the config path dynamically
based on the type of repo we are in.

c-alpha commented:
ping?

alexmv commented:
Apologies -- it's been a busy few weeks for me.

> > One option would be to have `dir_file` return `.git/config` if called without arguments, and if called with a path, call into `is_git_dir` to find an actual accurate path.
> 
> Why not just always call `is_git_dir` from `dir_file` then?

Because `is_git_dir` takes a path and actively examines its contents, and `dir_file` currently does not.
 
> I have experimented a bit, and it seems `is_git_dir` returns the directory where the git files are (i.e. `.../.git` for a working copy, the current directory for a bare repo, and `undef` outside a git tree). Thus, `/config` would need to be appended when the result is not `undef`.

Yup!

> This led me to the following in my script:
> 
> ```perl
> no warnings qw(redefine);
> sub Config::GitLike::Git::dir_file {
>     my $self = shift;
>     my $path = $self->is_git_dir(".");
>     if ((defined $path) and ($path ne "")) {
> 	my($vol, $dirs, undef) = File::Spec->splitpath( $path, 1 );
> 	my @dirs               = File::Spec->splitdir( $dirs );
> 	$path = File::Spec->catfile( @dirs, "config");
>     }
>     return $path;
> }
> use warnings qw(redefine);
> ```

Two minor comments: `$path` can't end up as `""` here, and one doesn't need to `splitpath`/`splitdir` before `catpath`'ing.

But I think the biggest problem with the above is that it always looks relative to `.`.  If run from inside a bare repo, you'll always get `config`, and if run from inside a non-bare repo, you'll get `.git/config` -- and if the CWD is _not_ a git repo, you'll get `undef`.  Depending on the CWD for this behavior (since the caller may be acting on a repo _anywhere_) seems like a bug.

How does the following work for you?  For your use case, you'd pass in to `dir_file` the some path you passed to `->load`:

```diff
diff --git lib/Config/GitLike.pm lib/Config/GitLike.pm
index ae8b545..202fe8f 100644
--- lib/Config/GitLike.pm
+++ lib/Config/GitLike.pm
@@ -1682,7 +1682,9 @@ precedence, you'll need to override L<"load"> as well.
 =head2 dir_file

 Return a string containing the path to a configuration file with the
-name C<confname> in a directory. The directory isn't specified here.
+name C<confname> in a directory.  Called with no arguments,
+returns the path for a generic directory; if called with a
+directory as an argument, returns the path for I<that> directory.

 =head2 global_file

diff --git lib/Config/GitLike/Git.pm lib/Config/GitLike/Git.pm
index 492195c..2169273 100644
--- lib/Config/GitLike/Git.pm
+++ lib/Config/GitLike/Git.pm
@@ -13,6 +13,17 @@ has '+compatible' => (
     default => 1,
 );

+sub dir_file {
+    my $self = shift;
+    return ".git/config" unless @_;
+    my $path = shift;
+    my $dir = $self->is_git_dir( $path );
+    return File::Spec->catfile( $dir, "config" ) if $dir;
+
+    $path = File::Spec->rel2abs( $path );
+    return File::Spec->catfile( $path, ".git/config");
+}
+
 sub is_git_dir {
     my $self = shift;
     my $path = File::Spec->rel2abs( shift );
@@ -83,7 +94,9 @@ This module overrides these methods from C<Config::GitLike>:

 =head2 dir_file

-The per-directory configuration file is F<.git/config>
+The per-directory configuration file is F<.git/config>.  With an
+optional directory argument, will return a fully-qualified path to the
+configuration file, as git would edit with C<git config --local -C path>.

 =head2 user_file

```

> If `dir_file` were changed to accept an argument, that would be an API change.

Sure -- but it's a backwards compatible one if it's an optional argument.

-----

It occurs to me that your problem may also be reasonably solved by writing to `$config->config_files->[-1]`.

c-alpha commented:
Sorry for the delay. I'll test your diff today!

c-alpha commented:
> [...]
> How does the following work for you? For your use case, you'd pass in to `dir_file` the some path you passed to `->load`:
> [...]

That works perfectly!

How do you want to handle it? I update the PR and you merge, or I close the PR as you'll be pushing it anyway?

alexmv commented:
Config::GitLike 1.18 is on its way to PAUSE.  Thanks for the bug report!

c-alpha commented:
> Config::GitLike 1.18 is on its way to PAUSE. Thanks for the bug report!

Many thanks for your support in resolving it!!! 👍
